lib/pure/strutils.nim

```diff
@@ -154,6 +154,124 @@ proc isDigit*(c: char): bool {.noSideEffect,
     doAssert isDigit('8') == true
   return c in Digits
 
+func isNumeric*(s: string, enableNaNInf, enableLooseDot = false): bool =
+  ## Checks whether the string is numeric.
+  ## When the string is an integer, float or exponential, it returns true,
+  ## otherwise it returns false.
+  ## The `enableNaNInf` value indicates whether the `NaN` and `Inf` values
+  ## are valid.
+  ## The `enableLooseDot` value indicates whether loose point values such as
+  ## `.9` and `9.` are valid. 
+  ##
+  ## **Note:** The reason that `parseFloat()` is not used to achieve this is its
+  ## poor performance.
+  runnableExamples:
+    doAssert isNumeric("123") == true
+    doAssert isNumeric("123.45") == true
+    doAssert isNumeric("+123.45") == true
+    doAssert isNumeric("123.45e-2") == true
+    doAssert isNumeric("+123.45E-2") == true
+    doAssert isNumeric("-123.45e2") == true
+    doAssert not isNumeric("e123.45") == true
+    doAssert not isNumeric("abc") == true
+    doAssert not isNumeric("123abc") == true
+    doAssert not isNumeric("123.45.6") == true
+    doAssert not isNumeric("123.45e++5") == true
+    doAssert not isNumeric("5.2+e1") == true
+    doAssert not isNumeric(".9") == true
+    doAssert not isNumeric("Inf") == true
+    doAssert not isNumeric("-Inf") == true
+    doAssert not isNumeric("NaN") == true
+    doAssert not isNumeric("nAn") == true
+    doAssert isNumeric("0.000_005") == true
+    doAssert isNumeric("1e1_00") == true
+    doAssert isNumeric("0.00_0001") == true
+    doAssert isNumeric("0.00__00_01") == true
+    doAssert isNumeric("0.0_01") == true
+    doAssert isNumeric("0.00_000_1") == true
+    doAssert isNumeric("0.00000_1") == true
+    doAssert isNumeric("1_0.00_0001") == true
+    doAssert isNumeric("1__00.00_0001") == true
+    doAssert not isNumeric("1_0._00_0001") == true
+    doAssert not isNumeric("_1_0_00.0001") == true
+    doAssert not isNumeric("10.00E_01") == true
+    doAssert not isNumeric("10.00E") == true
+    doAssert not isNumeric("10.00E+") == true
+    doAssert not isNumeric("10.00E-") == true
+    doAssert not isNumeric("10.00E_") == true
+    doAssert not isNumeric("10.00A") == true
+  let length = s.len
+  if length == 3:
+    if (s[0] in {'i', 'I'} and s[1] in {'n', 'N'} and s[2] in {'f', 'F'}) or
+       (s[0] in {'n', 'N'} and s[1] in {'a', 'A'} and s[2] in {'n', 'N'}):
+      if enableNaNInf:
+        return true
+      else:
+        return false
+  if length == 4:
+    if (s[0] in {'+', '-'} and s[1] in {'i', 'I'} and s[2] in {'n', 'N'} and s[3] in {'f', 'F'}) or
+       (s[0] in {'+', '-'} and s[1] in {'n', 'N'} and s[2] in {'a', 'A'} and s[3] in {'n', 'N'}):
+      if enableNaNInf:
+        return true
+      else:
+        return false
+
+  var i = 0
+  var eLeft, eRight, dot, e, num = false
+  while i < length:
```

Comment:
`for i, si in s: ...` leads to clearer code

Reply:
I've tested it multiple times on Windows and Linux, and the 'while' statement seems to be more efficient.

Reply:
that's troubling; do you have a reduced test case? because a lot of code relies on `for` being 0-cost compared to `while`

Reply:
may use .mitems or .mpairs for getting less cost when us `for`.

Reply:
Instead of a while statement, it's more efficient to use the form 'for i countup(0, sHigh)'.

Reply:
Nim compiles `for` to `while` anyways, so may not be the loop itself.

Reply:
this could be unrelated to while and instead related to:
```nim
for i, si in s:
  # use si
# vs
for i in low .. high:
  let si = s[i]
  # use si
```

but I don't understand the perf difference (not sure if related to pairs vs mpairs, etc)

Reply:
I think it is associated with pairs vs MPairs.

Reply:
it can't be, s is a let variable.

I still don't trust the benchmark you did (see above remarks).

these give identical times:

nim r -d:danger -d:case2a --gc:arc main
2.296005964279175
nim r -d:danger -d:case2b --gc:arc main
same + noise

```nim
when true:
  import times
  proc fn(s: string): int =
    var ret = 0
    when defined case2a:
      for i, si in s:
        ret += i
        ret += cast[int](si)
    when defined case2b:
      for i in 0..<s.len:
        ret += i
        ret += cast[int](s[i])
    return ret

  proc main =
    let ns = 1_000
    let n = 10_000_000
    var s: string
    for i in 0..<ns:
      s.add cast[char](i)
    var tot = 0
    for i in 0..<n:
      tot += fn(s)
    doAssert tot != -1

  let t = epochTime()
  main()
  echo epochTime() - t
```

Reply:
``` nim
import times

func isNumeric*(s: string, enableNaNInf, enableLooseDot = false): bool =
  ## Checks whether the string is numeric.
  ## When the string is an integer, float or exponential, it returns true,
  ## otherwise it returns false.
  ## The `enableNaNInf` value indicates whether the `NaN` and `Inf` values
  ## are valid.
  ## The `enableLooseDot` value indicates whether loose point values such as
  ## `.9` and `9.` are valid. 
  ##
  ## **Note:** The reason that `parseFloat()` is not used to achieve this is its
  ## poor performance.
  let length = s.len
  if length == 3:
    if (s[0] in {'i', 'I'} and s[1] in {'n', 'N'} and s[2] in {'f', 'F'}) or
       (s[0] in {'n', 'N'} and s[1] in {'a', 'A'} and s[2] in {'n', 'N'}):
      if enableNaNInf:
        return true
      else:
        return false
  if length == 4:
    if (s[0] in {'+', '-'} and s[1] in {'i', 'I'} and s[2] in {'n', 'N'} and s[3] in {'f', 'F'}) or
       (s[0] in {'+', '-'} and s[1] in {'n', 'N'} and s[2] in {'a', 'A'} and s[3] in {'n', 'N'}):
      if enableNaNInf:
        return true
      else:
        return false

  var eLeft, eRight, dot, e, num = false
  let sHigh = s.len - 1
  for i in countup(0, sHigh):
    case s[i]
    of '+', '-':
      if i == sHigh:
        return false
      if e == false:
        if num:
          return false
        if eLeft:
          return false
        eLeft = true
      else:
        if num:
          return false
        if eRight:
          return false
        eRight = true
    of '.':
      if dot:
        return false
      if not enableLooseDot:
        if num == false:
          return false
        else:
          if i == sHigh:
            return false
          if s[i+1] in {'e', 'E'}:
            return false
      num = false
      dot = true
    of 'e', 'E':
      if i == sHigh:
        return false
      if num or dot:
        if e:
          return false
        num = false
      else:
        return false
      e = true
    of '0'..'9':
      num = true
    of '_':
      if num == false:
        return false
      if dot and num == false:
        return false
      if e and num == false:
        return false
    else:
      return false

  return true

func isNumeric2*(s: string, enableNaNInf, enableLooseDot = false): bool =
  ## Checks whether the string is numeric.
  ## When the string is an integer, float or exponential, it returns true,
  ## otherwise it returns false.
  ## The `enableNaNInf` value indicates whether the `NaN` and `Inf` values
  ## are valid.
  ## The `enableLooseDot` value indicates whether loose point values such as
  ## `.9` and `9.` are valid. 
  ##
  ## **Note:** The reason that `parseFloat()` is not used to achieve this is its
  ## poor performance.
  let length = s.len
  if length == 3:
    if (s[0] in {'i', 'I'} and s[1] in {'n', 'N'} and s[2] in {'f', 'F'}) or
       (s[0] in {'n', 'N'} and s[1] in {'a', 'A'} and s[2] in {'n', 'N'}):
      if enableNaNInf:
        return true
      else:
        return false
  if length == 4:
    if (s[0] in {'+', '-'} and s[1] in {'i', 'I'} and s[2] in {'n', 'N'} and s[3] in {'f', 'F'}) or
       (s[0] in {'+', '-'} and s[1] in {'n', 'N'} and s[2] in {'a', 'A'} and s[3] in {'n', 'N'}):
      if enableNaNInf:
        return true
      else:
        return false

  var eLeft, eRight, dot, e, num = false
  let sHigh = s.len - 1
  for i, si in s:
    case si
    of '+', '-':
      if i == sHigh:
        return false
      if e == false:
        if num:
          return false
        if eLeft:
          return false
        eLeft = true
      else:
        if num:
          return false
        if eRight:
          return false
        eRight = true
    of '.':
      if dot:
        return false
      if not enableLooseDot:
        if num == false:
          return false
        else:
          if i == sHigh:
            return false
          if s[i+1] in {'e', 'E'}:
            return false
      num = false
      dot = true
    of 'e', 'E':
      if i == sHigh:
        return false
      if num or dot:
        if e:
          return false
        num = false
      else:
        return false
      e = true
    of '0'..'9':
      num = true
    of '_':
      if num == false:
        return false
      if dot and num == false:
        return false
      if e and num == false:
        return false
    else:
      return false

  return true

proc testIsNumeric() =
  var a = 0
  let execptedValue = 160000000
  let t0 = cputime()
  for i in 0..9999999:
    if isNumeric("123"): a += 1
    if isNumeric("123.45"): a += 1
    if isNumeric("+123.45"): a += 1
    if isNumeric("123.45e-2"): a += 1
    if isNumeric("+123.45E-2"): a += 1
    if isNumeric("-123.45e2"): a += 1
    if not isNumeric("e123.45"): a += 1
    if not isNumeric("abc"): a += 1
    if not isNumeric("123abc"): a += 1
    if not isNumeric("123.45.6"): a += 1
    if not isNumeric("123.45e++5"): a += 1
    if not isNumeric("5.2+e1"): a += 1
    if not isNumeric(".9"): a += 1
    if not isNumeric("Inf"): a += 1
    if not isNumeric("-Inf"): a += 1
    if not isNumeric("NaN"): a += 1
  let t1 = cputime() - t0
  if a == execptedValue:
    echo "OK! isNumeric() time consuming:", t1

proc testIsNumeric2() =
  var a = 0
  let execptedValue = 160000000
  let t0 = cputime()
  for i in 0..9999999:
    if isNumeric2("123"): a += 1
    if isNumeric2("123.45"): a += 1
    if isNumeric2("+123.45"): a += 1
    if isNumeric2("123.45e-2"): a += 1
    if isNumeric2("+123.45E-2"): a += 1
    if isNumeric2("-123.45e2"): a += 1
    if not isNumeric2("e123.45"): a += 1
    if not isNumeric2("abc"): a += 1
    if not isNumeric2("123abc"): a += 1
    if not isNumeric2("123.45.6"): a += 1
    if not isNumeric2("123.45e++5"): a += 1
    if not isNumeric2("5.2+e1"): a += 1
    if not isNumeric2(".9"): a += 1
    if not isNumeric2("Inf"): a += 1
    if not isNumeric2("-Inf"): a += 1
    if not isNumeric2("NaN"): a += 1
  let t1 = cputime() - t0
  if a == execptedValue:
    echo "OK! isNumeric2() time consuming:", t1

testIsNumeric()
testIsNumeric2()
```
The time difference is small, but there is still a difference.
OK! isNumeric() time consuming:1.82
OK! isNumeric2() time consuming:2.25

Reply:
-d:release
OK! isNumeric() time consuming:2.303698
OK! isNumeric2() time consuming:2.077815


-d:danger
OK! isNumeric() time consuming:1.999999999999832e-06
OK! isNumeric2() time consuming:1.999999999999398e-06

only -d:danger is relevant for benchmarkings, not -d:release

so there is no difference AFAIK, and the simplest form should be used (`or i, si in s:`)